google/cloud/forseti/enforcer/gce_firewall_enforcer.py

```diff
@@ -923,16 +933,44 @@ def _update_rules(self, network):
                 self.expected_rules.rules[rule_name]
                 for rule_name in self._rules_to_update
             ], network)
-            patch_function = self.compute_client.patch_firewall_rule
-            (successes, failures, change_errors) = self._apply_change(
-                patch_function, rules)
-            self._updated_rules.extend(successes)
-            change_count += len(successes)
-            if failures:
-                raise FirewallEnforcementUpdateFailedError(
-                    'Firewall enforcement failed while deleting rules for '
-                    'project {}. The following errors were encountered: {}'
+            rules_to_patch = []
+            rules_to_replace = []
+            for rule in rules:
+                if _rule_update_can_patch(self.current_rules.rules[rule['name']], rule):
+                    rules_to_patch.append(rule)
+                else:
+                    rules_to_replace.append(rule)
+            if rules_to_patch:
+                change_count += self._patch_rules(rules_to_patch)
+            if rules_to_replace:
+                change_count += self._replace_rules(rules_to_replace)
+        return change_count
```

Comment:
Perhaps you can add a new unit test that would test the replace logic is working? Maybe something like [this](https://github.com/forseti-security/forseti-security/blob/c2ae170a4b9ce19208c054ba51f97e154dfba247/tests/enforcer/gce_firewall_enforcer_test.py#L995)?

Reply:
captured in test_apply_firewall_denied_to_allowed

Since the replacement rules are captured as 'updates' since they share the same name. replace vs update is a deeper implementation decision, so test_apply_firewall_denied_to_allowed captures the full stack to confirm these rules update.